crates/rover-client/src/operations/persisted_queries/publish/types.rs

```diff
@@ -107,6 +110,87 @@ impl From<PersistedQueriesPublishInput> for QueryVariables {
+        let mut anonymous_operations = Vec::new();
+        let mut operations = Vec::new();
+        for (id, body) in relay_manifest.operations {
+            let ast = Parser::new(&body).parse();
+
+            for definition in ast.document().definitions() {
+                if let Definition::OperationDefinition(operation_definition) = definition {
+                    // attempt to extract operation name, tracking IDs without them
+                    let operation_name = if let Some(name) = operation_definition.name() {
+                        Some(name)
+                    } else {
+                        anonymous_operations.push(id.clone());
+                        None
+                    };
+
+                    // attempt to extract operation type, tracking IDs without them
+                    let operation_type = if let Some(ty) = operation_definition.operation_type() {
+                        Some(ty)
+                    } else {
+                        operations_with_invalid_type.push(id.clone());
+                        None
+                    };
+
+                    // if both a name and a type were found, record them
+                    if let (Some(operation_name), Some(operation_type)) =
+                        (operation_name, operation_type)
+                    {
+                        operations.push(PersistedQueryOperation {
+                            name: operation_name.text().to_string(),
+                            r#type: match (operation_type.mutation_token(), operation_type.query_token(), operation_type.subscription_token()) {
+                                (Some(_mutation), _, _) => PersistedQueryOperationType::Mutation,
+                                (_, Some(_query), _) => PersistedQueryOperationType::Query,
+                                (_, _, Some(_subscription)) => PersistedQueryOperationType::Subscription,
+                                _ => unreachable!("operation type is guaranteed to be one of query, mutation, or subscription")
+                            },
+                            body: body.to_string(),
+                            id: id.to_string()
+                        });
+                    }
+                }
+            }
+        }
+
+        let mut errors = Vec::new();
+
+        if !anonymous_operations.is_empty() {
+            errors.push(format!(
+                "The following operation IDs do not have a name: {}.",
+                anonymous_operations.join(", ")
+            ));
+        }
+
+        if !operations_with_invalid_type.is_empty() {
+            errors.push(format!("The following operation IDs do not have a valid operation type: {}, must be one of 'query', 'mutation', or 'subscription'.", operations_with_invalid_type.join(", ")));
+        }
+
+        if errors.is_empty() {
+            let manifest = ApolloPersistedQueryManifest { operations };
+            if let Ok(json) = serde_json::to_string(&manifest) {
```

Comment:
What is this line for? It looks like we're serializing it just for the sake of checking that there are no errors, and throwing away the result? Won't we get this error later anyway (when we try to send the RPC) in the off chance it's possible?